File: FBDeviceControlTests/Tests/Unit/FBAMDServiceConnectionReaderTests.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBControlCore/FBControlCore.h>
#import <FBDeviceControl/FBDeviceControl-Swift.h>

static AMSecureIOContext NullSecureIOContext(CFTypeRef connection)
{
  return NULL;
}

static int32_t EndOfFileReceive(CFTypeRef connection, void *buffer, size_t bytes)
{
  return 0;
}

// Gate for the blocking-receive fake: receive blocks until invalidation
// signals it, mimicking a socket read unblocked by connection invalidation.
static dispatch_semaphore_t sReceiveGate;

static int32_t BlockingReceive(CFTypeRef connection, void *buffer, size_t bytes)
{
  dispatch_semaphore_wait(sReceiveGate, DISPATCH_TIME_FOREVER);
  return 0;
}

static int InvalidateUnblockingReceive(CFTypeRef connection)
{
  dispatch_semaphore_signal(sReceiveGate);
  return 0;
}

@interface FBAMDServiceConnectionReaderTests : XCTestCase

@end

@implementation FBAMDServiceConnectionReaderTests

- (FBAMDServiceConnection *)connectionReceivingEndOfFile
{
  AMDCalls calls = {};
  calls.ServiceConnectionGetSecureIOContext = NullSecureIOContext;
  calls.ServiceConnectionReceive = EndOfFileReceive;
  return [FBAMDServiceConnection
          connectionWithName:@"test-connection"
          connection:(__bridge AMDServiceConnectionRef) @"fake-connection-ref"
          device:NULL
          calls:calls
          logger:FBControlCoreGlobalConfiguration.defaultLogger];
}

- (void)testReaderDeliversEndOfFileToTheConsumer
{
  FBAMDServiceConnection *connection = [self connectionReceivingEndOfFile];
  id<FBAccumulatingBuffer> consumer = FBDataBuffer.accumulatingBuffer;
  dispatch_queue_t queue = dispatch_queue_create("com.facebook.fbdevicecontrol.tests.reader", DISPATCH_QUEUE_SERIAL);
  id<FBFileReaderProtocol> reader = [connection readFromConnectionWritingToConsumer:consumer onQueue:queue];

  NSError *error = nil;
  XCTAssertNotNil([reader.startReading awaitWithTimeout:5 error:&error]);
  XCTAssertNotNil([consumer.finishedConsuming awaitWithTimeout:5 error:&error]);

  NSNumber *finished = [reader.finishedReading awaitWithTimeout:5 error:&error];
  XCTAssertEqualObjects(finished, @(FBFileReaderStateFinishedReadingNormally));
}

- (void)testInvalidationWaitsForInFlightReadToFinish
{
  sReceiveGate = dispatch_semaphore_create(0);
  AMDCalls calls = {};
  calls.ServiceConnectionGetSecureIOContext = NullSecureIOContext;
  calls.ServiceConnectionReceive = BlockingReceive;
  calls.ServiceConnectionInvalidate = InvalidateUnblockingReceive;
  FBAMDServiceConnection *connection = [FBAMDServiceConnection
                                        connectionWithName:@"test-connection"
                                        connection:(__bridge AMDServiceConnectionRef) @"fake-connection-ref"
                                        device:NULL
                                        calls:calls
                                        logger:FBControlCoreGlobalConfiguration.defaultLogger];
  id<FBAccumulatingBuffer> consumer = FBDataBuffer.accumulatingBuffer;
  dispatch_queue_t queue = dispatch_queue_create("com.facebook.fbdevicecontrol.tests.blocked-reader", DISPATCH_QUEUE_SERIAL);
  id<FBFileReaderProtocol> reader = [connection readFromConnectionWritingToConsumer:consumer onQueue:queue];

  NSError *error = nil;
  XCTAssertNotNil([reader.startReading awaitWithTimeout:5 error:&error]);

  // Invalidation unblocks the (gated) receive and must not return until the
  // read loop has exited.
  XCTAssertTrue([connection invalidateWithError:&error]);
  XCTAssertEqual(reader.finishedReading.state, FBFutureStateDone);
}

@end

File: FBDeviceControlTests/Tests/Unit/FBAMDServiceConnectionReaderTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
@testable import FBDeviceControl
import Foundation
import Testing

// MARK: - File-scope state for C function pointer callbacks

// Gate for the blocking-receive fake: receive blocks until invalidation signals it, mimicking a
// socket read unblocked by connection invalidation. File-scope because a C function pointer
// cannot capture; the suite is serialized so tests cannot race on it.
private var sReceiveGate = DispatchSemaphore(value: 0)

private func endOfFileReceive(_ connection: CFTypeRef?, _ buffer: UnsafeMutableRawPointer?, _ bytes: Int) -> Int32 {
  0
}

private func blockingReceive(_ connection: CFTypeRef?, _ buffer: UnsafeMutableRawPointer?, _ bytes: Int) -> Int32 {
  sReceiveGate.wait()
  return 0
}

private func invalidateUnblockingReceive(_ connection: CFTypeRef?) -> Int32 {
  sReceiveGate.signal()
  return 0
}

@Suite(.serialized)
struct FBAMDServiceConnectionReaderTests {

  private func makeConnection(calls: AMDCalls, connection: AnyObject = "fake-connection-ref" as AnyObject) -> FBAMDServiceConnection {
    FBAMDServiceConnection(
      name: "test-connection",
      connection: connection,
      device: "fake-device-ref" as AnyObject,
      calls: calls,
      logger: FBControlCoreGlobalConfiguration.defaultLogger)
  }

  @Test
  func readerDeliversEndOfFileToTheConsumer() async throws {
    var calls = FBCreateZeroedAMDCalls()
    calls.ServiceConnectionGetSecureIOContext = { _ in nil }
    calls.ServiceConnectionReceive = endOfFileReceive
    let connection = makeConnection(calls: calls)
    let consumer = FBDataBuffer.accumulatingBuffer()
    let queue = DispatchQueue(label: "com.facebook.fbdevicecontrol.tests.reader")
    let reader = connection.readFromConnectionWriting(to: consumer, on: queue)

    _ = try await bridgeFBFuture(reader.startReading())
    _ = try await bridgeFBFuture(consumer.finishedConsuming)

    let finished = try await bridgeFBFuture(reader.finishedReading)
    #expect(finished == NSNumber(value: FBFileReaderState.finishedReadingNormally.rawValue))
  }

  @Test
  func invalidationWaitsForInFlightReadToFinish() async throws {
    sReceiveGate = DispatchSemaphore(value: 0)
    var calls = FBCreateZeroedAMDCalls()
    calls.ServiceConnectionGetSecureIOContext = { _ in nil }
    calls.ServiceConnectionReceive = blockingReceive
    calls.ServiceConnectionInvalidate = invalidateUnblockingReceive
    // The reference is held here so the retain below lands on the same instance the connection
    // stores: invalidate releases it, standing in for the +1 MobileDevice hands over.
    let connectionRef = "fake-connection-ref" as AnyObject
    let connection = makeConnection(calls: calls, connection: connectionRef)
    _ = Unmanaged.passRetained(connectionRef)
    let consumer = FBDataBuffer.accumulatingBuffer()
    let queue = DispatchQueue(label: "com.facebook.fbdevicecontrol.tests.blocked-reader")
    let reader = connection.readFromConnectionWriting(to: consumer, on: queue)

    _ = try await bridgeFBFuture(reader.startReading())

    // Invalidation unblocks the (gated) receive and must not return until the read loop has
    // exited.
    try connection.invalidate()
    #expect(reader.finishedReading.state == .done)
  }
}
